### PR TITLE
Introduction of seperate Consumer class and change in the polling strategy

### PR DESCRIPTION
**Pull request by elpoelma**
This PR introduces the following major functionality:

- A new Consumer class which  wraps the functionality of https://github.com/TREEcg/event-stream-client/tree/main/packages/actor-init-ldes-client and allows you to provide a callback which is called on each incoming member of the stream.
- The consumer now runs at a specific interval determined by a cron pattern.
- Aside from the last seen timestamp, the last seen page is now transmitted as well.
- When processing the members, blank nodes are replace by named nodes based on a uuid.

### CODE REVIEW

File: app.ts [hunk 1]
```diff
@@ -1,96 +1,52 @@
...
+		quadsToRemove.push(
+			quad(variable("s"), PURL("isVersionOf"), baseResourceUri)
+		);
+    quadsToRemove.push(quad(variable("s"), variable("p"), variable("o")))
+	}
+	await executeDeleteInsertQuery(quadsToRemove, quadsToAdd);
 }
 
-main();
+const consumerJob = new CronJob(CRON_PATTERN, async () => {
```
Comment (cecemel) [inline app.ts:31]: Not sure if the cron library handles this, but suppose long running job, do we run into the risk of having two executions at the same time?

File: utils.ts [hunk 2]
```diff
@@ -1,20 +1,74 @@
...
-    default:
-      return term.value;
-  }
+	switch (term.termType) {
+		case "NamedNode":
+			return `<${term.value}>`;
+		case "Literal":
+			let result = `"${term.value}"`;
+			if (term.datatype) {
+				result += `^^${toString(term.datatype)}`;
```
Comment (cecemel) [inline utils.ts]: Can you also pick up langstrings?
See eg: https://github.com/lblod/lpdc-management-service/blob/master/utils/bindingsToNT.js#L17

File: utils.ts [hunk 3]
```diff
@@ -1,20 +1,74 @@
...
+			}
+			quad.subject = blankNodesMap.get(quad.subject)!;
+		}
+		if (quad.object.termType === "BlankNode") {
+			if (!blankNodesMap.has(quad.object)) {
+				blankNodesMap.set(quad.object, BLANK(uuidv4()));
+			}
+			quad.object = blankNodesMap.get(quad.object)!;
+		}
+	});
```
Comment (cecemel) [inline utils.ts]: I generally like returns to these kind of utils. But you do changes in place, so strictly speaking no needed.

File: consumer.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+			representation: "Quads",
+			mimeType: "application/ld+json",
+			emitMemberOnce: true,
+			fromTime: this.startTimeStamp,
+			disableSynchronization: true, 
+		});
+		stream.on("metadata", (metadata) => {
+			this.currentPage = metadata.url;
+		}); 
+		stream.on("data", async (member: Member) => {
```
Comment (cecemel) [inline consumer.ts:53]: so  .on('error')?

File: README.md [hunk 5]
```diff
@@ -1,16 +1,33 @@
...
+
+The service can be configured with the following environment variables:
+
+-   `LDES_ENDPOINT_VIEW` [string]: the ldes endpoint containing the view (the first page) of the stream.
+
+-   `LDES_STREAM` [string]: the uri which should be used as a subject to store the latest page and timestamp consumed in the database. (default: `http://mu.semte.ch/example-stream`)
+
+-   `LDES_RELATION_PATH` [string]: the predicate which is being used to define relations. For time-based streams, this will almost always be http://www.w3.org/ns/prov#generatedAtTime. (default: `http://www.w3.org/ns/prov#generatedAtTime`)
+
+-   `LDES_POLLING_INTERVAL` [number]: the polling interval in milliseconds of the client which fetches resources from and endpoint (default: `500`)
```
Comment (cecemel) [inline README.md]: still needed here?

File: sparql-queries.ts [hunk 6]
```diff
@@ -1,148 +1,165 @@
...
-  let queryStr = constructInsertQuery(quads);
-  try {
-    await update(queryStr);
-  } catch (e) {
-    console.error(e);
-  }
+export async function executeInsertQuery(quads: RDF.Quad[]) {
+	let queryStr = constructInsertQuery(quads);
+	try {
+		await update(queryStr);
```
Comment (cecemel) [inline sparql-queries.ts]: tabs vs spaces?

File: utils.ts (hunk 2 above)
Comment (cecemel) [reply utils.ts]: Btw I see a potential bug in my example, you should first check for lang and then datatype. Syntax rdf in langstrings is really weird

File: utils.ts [hunk 7]
```diff
@@ -1,20 +1,74 @@
...
+		case "Literal":
+			let result = `"${term.value}"`;
+			if (term.datatype) {
+				result += `^^${toString(term.datatype)}`;
+			}
+			return result;
+		case "Quad":
+			return `${toString(term.subject)} ${toString(
+				term.predicate
+			)} ${toString(term.object)}.`;
```
Comment (cecemel) [inline utils.ts]: could you also use the sparqlEscape helpers? cause it has specific escaping rules.

**Review by cecemel — CHANGES_REQUESTED** [review]

File: app.ts [hunk 8]
```diff
@@ -1,96 +1,52 @@
...
+	const quadsToAdd: RDF.Quad[] = member.quads;
+	const quadsToRemove: RDF.Quad[] = [];
+	const baseResourceUri = extractBaseResourceUri(member);
+	if (baseResourceUri && REPLACE_VERSIONS) {
+		quadsToRemove.push(
+			quad(variable("s"), PURL("isVersionOf"), baseResourceUri)
+		);
+    quadsToRemove.push(quad(variable("s"), variable("p"), variable("o")))
+	}
+	await executeDeleteInsertQuery(quadsToRemove, quadsToAdd);
```
Comment (cecemel) [inline app.ts]: I am not sure how we should use this callback function to execute custom storage rules?

Comment (cecemel) [reply app.ts]: my idea ideally would be: you mount a custom file in /config with specifc signature, so you could also I dunno do some other calls.
What we have as default behaviour is nice btw. But in the context of ipd-lpdc, we'll have to write a consumer that pushes a json ld call

**Review by cecemel — CHANGES_REQUESTED** [review]

File: app.ts (hunk 8 above)
Comment (elpoelma) [reply app.ts]: The idea would be that we extract the Consumer class to a seperate library, which can also be used for a service which pushes json calls.

Comment (cecemel) [reply app.ts]: ah ok

**Pull request merged**